### PR TITLE
Conio : Fix bug and textcolor/bgcolor added

### PR DESCRIPTION
**Pull request by jedeoric**
Here is a PR in order to update conio :

* Fix a bug for gotoxy : Y did not work because it required to update screen address.
* Add bgcolor and textcolor management. 

Please read telestrat doc update in order to see limits with bgcolor and textcolor.

### CODE REVIEW

File: libsrc/telestrat/gotoxy.s [hunk 1]
```diff
@@ -5,10 +5,9 @@
 ; void gotoxy (unsigned char x, unsigned char y);
 ;
 
-        .export         gotoxy, _gotoxy
+        .export         gotoxy, _gotoxy, _update_adscr
```
Comment (greg-king5) [inline libsrc/telestrat/gotoxy.s]: Do you want `update_adscr()` to be used by C code?  Does the name need that leading underscore?

File: libsrc/telestrat/bgcolor.s [hunk 2]
```diff
@@ -0,0 +1,28 @@
+; 2019-07-02, Jede (jede@oric.org)
+;
+
+        .export         _bgcolor
+        .import         BGCOLOR
+        .import         BGCOLOR_CHANGE
+        .include        "telestrat.inc"
+
+.proc _bgcolor
+    cmp     BGCOLOR       ; Do we set the same color? if we don't detect it, we loose one char on the screen for each textcolor call with the same color
```
Comment (oliverschmidt) [inline libsrc/telestrat/bgcolor.s]: Upper case after '?'.

File: libsrc/telestrat/textcolor.s [hunk 3]
```diff
@@ -0,0 +1,28 @@
+; 2019-07-02, Jede (jede@oric.org)
+;
+
+        .export         _textcolor
+        .import         CHARCOLOR
+        .import         CHARCOLOR_CHANGE
+        .include        "telestrat.inc"
+
+.proc _textcolor
+    cmp     CHARCOLOR       ; Do we set the same color? if we don't detect it, we loose one char on the screen for each textcolor call with the same color
```
Comment (oliverschmidt) [inline libsrc/telestrat/textcolor.s]: Upper case after '?'.

File: libsrc/telestrat/textcolor.s [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    bne     out             ; yes
+    lda     #$00
+    sta     CHARCOLOR_CHANGE
+    
+    lda     CHARCOLOR       ; return last color
```
Comment (oliverschmidt) [inline libsrc/telestrat/textcolor.s]: Upper case after ';'.

File: libsrc/telestrat/textcolor.s [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+;
+
+        .export         _textcolor
+        .import         CHARCOLOR
+        .import         CHARCOLOR_CHANGE
+        .include        "telestrat.inc"
+
+.proc _textcolor
+    cmp     CHARCOLOR       ; Do we set the same color? if we don't detect it, we loose one char on the screen for each textcolor call with the same color
+    bne     out             ; yes
```
Comment (oliverschmidt) [inline libsrc/telestrat/textcolor.s]: Upper case after ';'.

File: libsrc/telestrat/textcolor.s [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    lda     #$00
+    sta     CHARCOLOR_CHANGE
+    
+    lda     CHARCOLOR       ; return last color
+
+    rts
+out:
+    ldx     CHARCOLOR       ; get last color in order to return it
```
Comment (oliverschmidt) [inline libsrc/telestrat/textcolor.s]: Upper case after ';'.

File: libsrc/telestrat/textcolor.s [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    sta     CHARCOLOR
+
+    lda     #$01
+    sta     CHARCOLOR_CHANGE
+    txa                     ; return previous color
```
Comment (oliverschmidt) [inline libsrc/telestrat/textcolor.s]: Upper case after ';'.

File: libsrc/telestrat/textcolor.s [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    rts
+.endproc
+
```
Comment (oliverschmidt) [inline libsrc/telestrat/textcolor.s]: Please remove trailing empty line.

File: libsrc/telestrat/gotoxy.s [hunk 9]
```diff
@@ -18,9 +17,38 @@ gotoxy: jsr     popa            ; Get Y
...
+        adc     #$28
+        bcc     skip
+        inc     ADSCRH
+skip:
+        sta     ADSCRL        
+        dey
+        bne     loop
+out:        
+        rts
+.endproc
```
Comment (oliverschmidt) [inline libsrc/telestrat/gotoxy.s]: Please add missing final newline.

File: libsrc/telestrat/bgcolor.s [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+out:
+    ldx     BGCOLOR       ; Get last color in order to return it
+    sta     BGCOLOR
+
+    lda     #$01          ; Notify the change color  
+    sta     BGCOLOR_CHANGE  
+    txa                   ; Return previous color
+    rts
+.endproc
+
```
Comment (oliverschmidt) [inline libsrc/telestrat/bgcolor.s]: Please remove trailing blank lines.

File: libsrc/telestrat/cputc.s [hunk 11]
```diff
@@ -4,12 +4,43 @@
...
+    sta     BGCOLOR_CHANGE
+
+    pla
+
+do_not_change_color:
+    BRK_TELEMON  XFWR            ; Macro send char to screen (channel 0)
     rts
 .endproc
-
+CHARCOLOR:
```
Comment (oliverschmidt) [inline libsrc/telestrat/cputc.s]: These variables very likely belong into the BSS.

File: libsrc/telestrat/bgcolor.s [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        .import         BGCOLOR_CHANGE
+        .include        "telestrat.inc"
+
+.proc _bgcolor
+    cmp     BGCOLOR       ; Do we set the same color? if we don't detect it, we loose one char on the screen for each textcolor call with the same color
+    bne     out           ; Yes
+    lda     #$00
+    sta     BGCOLOR_CHANGE
+    
+    lda     BGCOLOR       ; Return last color
```
[the remark this answers is not in the document]
Comment (jedeoric) [reply libsrc/telestrat/bgcolor.s]: Fixed thanks for the optimization

File: libsrc/telestrat/gotoxy.s (hunk 1 above)
Comment (jedeoric) [reply libsrc/telestrat/gotoxy.s]: No, update_adscr is not available from C.  I corrected the mistake. Fixed. Thanks.

File: libsrc/telestrat/textcolor.s (hunk 3 above)
Comment (jedeoric) [reply libsrc/telestrat/textcolor.s]: Fixed

File: libsrc/telestrat/textcolor.s (hunk 5 above)
Comment (jedeoric) [reply libsrc/telestrat/textcolor.s]: Fixed

File: libsrc/telestrat/textcolor.s (hunk 6 above)
Comment (jedeoric) [reply libsrc/telestrat/textcolor.s]: Fixed

File: libsrc/telestrat/textcolor.s (hunk 4 above)
Comment (jedeoric) [reply libsrc/telestrat/textcolor.s]: Fixed

File: libsrc/telestrat/textcolor.s (hunk 7 above)
Comment (jedeoric) [reply libsrc/telestrat/textcolor.s]: Fixed

File: libsrc/telestrat/gotoxy.s (hunk 9 above)
Comment (jedeoric) [reply libsrc/telestrat/gotoxy.s]: Fixed

File: libsrc/telestrat/bgcolor.s (hunk 10 above)
Comment (jedeoric) [reply libsrc/telestrat/bgcolor.s]: Fixed

File: libsrc/telestrat/textcolor.s (hunk 8 above)
Comment (jedeoric) [reply libsrc/telestrat/textcolor.s]: Fixed

File: libsrc/telestrat/bgcolor.s (hunk 2 above)
Comment (jedeoric) [reply libsrc/telestrat/bgcolor.s]: Fixed

File: libsrc/telestrat/cputc.s (hunk 11 above)
Comment (jedeoric) [reply libsrc/telestrat/cputc.s]: Done, moved to bss segment.

Thank a lot for the review

**Comment by jedeoric** [comment]
Corrections done

File: libsrc/telestrat/textcolor.s [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+        .include        "telestrat.inc"
+
+.proc _textcolor
+    cmp     CHARCOLOR       ; Do we set the same color? If we don't detect it, we loose one char on the screen for each textcolor call with the same color
+    bne     out             ; yes
+    
+    ldy     #$00
+    sty     CHARCOLOR_CHANGE
+    
+    lda     CHARCOLOR       ; Return last color
```
Comment (greg-king5) [inline libsrc/telestrat/textcolor.s]: The two lines above are not needed (similar to `bgcolor.s`).

File: libsrc/telestrat/cputc.s [hunk 14]
```diff
@@ -4,12 +4,46 @@
...
 
 .proc _cputc
-    BRK_TELEMON  XWR0       ; macro send char to screen (channel 0 in telemon terms)
+    ldx     CHARCOLOR_CHANGE
+    beq     do_not_change_color_foreground
+    dec     SCRX
+    dec     SCRX
+    pha
+    lda     CHARCOLOR
+    BRK_TELEMON  $4E             ; Change color on the screen (foreground)
```
Comment (greg-king5) [inline libsrc/telestrat/cputc.s]: The macro argument should be `XFWR` instead of `$4E`.

File: libsrc/telestrat/clrscr.s [hunk 15]
```diff
@@ -34,5 +34,14 @@
     sta     SCRY
     lda     #$00
     sta     SCRX
+
+    lda     #$00
```
Comment (greg-king5) [inline libsrc/telestrat/clrscr.s]: If you use an index register, then you can put `$01` into `SCRY`, decrement it, then put zero into the other three variables.

**Comment by polluks** [comment]
You could place the two flags in one byte and use the BIT instruction...

**Comment by greg-king5** [comment]
Also, you can make `update_adscr` set the color change flags.  Then, `gotoy()` wouldn't need to do it.

**Comment by jedeoric** [comment]
I need to have a look. Please wait, i need some time to fix it

File: libsrc/telestrat/textcolor.s (hunk 13 above)
Comment (jedeoric) [reply libsrc/telestrat/textcolor.s]: Corrected

File: libsrc/telestrat/clrscr.s (hunk 15 above)
Comment (jedeoric) [reply libsrc/telestrat/clrscr.s]: Done

**Comment by jedeoric** [comment]
> 
> 
> Also, you can make `update_adscr` set the color change flags. Then, `gotoy()` wouldn't need to do it.

Done

**Comment by jedeoric** [comment]
> 
> 
> You could place the two flags in one byte and use the BIT instruction...

I can but it means that the optimization in bgcolor.s and bgcolor.s are lost :

In update_adscr : 

Instead of 
        lda     #$01
        sta     CHARCOLOR_CHANGE
        sta     BGCOLOR_CHANGE

i can write 

        lda     #%11000000                                ; bit 6 and bit 7 are set to 1 (bit 6 : textcolor color, bit 7 bgcolor)
        sta     CHARCOLOR_CHANGE
        ; we earn 3 bytes (sta $xxxx removed)

But in textcolor.s : 
instead of line 13 : 
   ; 5 bytes
    ldy     #$00
    sty     CHARCOLOR_CHANGE

we wrote instead : 
   ; 9 bytes
    tax ; save old color         lost 1 byte
    lda CHARCOLOR_CHANGE
    ora #%0100000            ; lost 2 bytes
    sta    CHARCOLOR_CHANGE
    txa ; load old color         ; lost 1 byte

Still in textcolor.s, line 23, instead of 
   ; 5 bytes
    lda     #$01
    sta     CHARCOLOR_CHANGE

I should write 
   ;  8 bytes
   lda     CHARCOLOR_CHANGE  ; lost 1 byte
   ora     #%01000000                 ; lost 2 bytes
   sta     CHARCOLOR_CHANGE


It's the same for bgcolor.s (for bit 7) which we lost at least 7 bytes if we repeat the same code

In cputc.s, i think that we could lost some bytes too, i did a test, but i stopped because i think that we lost too much bytes and cycles in bgcolor and textcolor.

Let me know if you want me to change the code with color change in one byte, but i think that we lost bytes and cycles.

**Comment by oliverschmidt** [comment]
@polluks: As far as I see this thread waits for your comment...

**Comment by polluks** [comment]
@oliverschmidt It's ok :)

**Comment by greg-king5** [comment]
I just realized that this PR's way of handling color changes might fail.  Consider this program flow:

- `textcolor()` is called with a new color.  A character is printed (in the new color).
- `textcolor()` is called a second time with a different color.
- `textcolor()` is called a third time before any characters are printed.

If the third color is the same as the second color, then the "color_change" flag would be cleared; the second color wouldn't be printed.

But, if the third color is the same as the first color, then a second attribute would be put into screen RAM -- even though it wouldn't be needed.

---
I think that you need to use a "new color -- old color" method:  The color functions would set "new color" variables.  `cputc()` would set an attribute if the new color is different from the old one; then, it would copy the new color into the "old color" variable.  `clrscr()` and `update_adscr` would set the "old color" variables to the hardware defaults.

**Comment by jedeoric** [comment]
Good point, i try to correct it as soon as possible

**Comment by jedeoric** [comment]
Fixed @greg-king5 

Sorry for the delay, i was a bit busy and i had a bug. textcolor, bgcolor and cputc use old_color to set or not attributes

**Comment by jedeoric** [comment]
@polluks  Fixed

File: libsrc/telestrat/gotoxy.s [hunk 16]
```diff
@@ -19,8 +18,43 @@ gotoxy: jsr     popa            ; Get Y
...
-        jsr     popa
-        sta     SCRX
-        rts
+    sta     SCRY
+    
+
+    jsr     popa
+    sta     SCRX
+
+    jmp     update_adscr          ; Update adress video ram position when SCRY is modified
```
Comment (greg-king5) [inline libsrc/telestrat/gotoxy.s]: That line jumps to the next instruction.  You should comment it out.
And, add a comment on a line below it that says something like,
```asm
; Fall through
```

File: libsrc/telestrat/gotoxy.s [hunk 17]
```diff
@@ -19,8 +18,43 @@ gotoxy: jsr     popa            ; Get Y
...
+
+    lda     #>SCREEN
+    sta     ADSCRH
+
+    ldy     SCRY
+    beq     out
+loop:
+    lda     ADSCRL          
+    clc
+    adc     #$28
```
Comment (greg-king5) [inline libsrc/telestrat/gotoxy.s]: You should change `$28` to `SCREEN_XSIZE`.

File: libsrc/telestrat/bgcolor.s [hunk 18]
```diff
@@ -3,23 +3,10 @@
...
-    cmp     BGCOLOR       ; Do we set the same color? If we don't detect it, we loose one char on the screen for each bgcolor call with the same color
-    bne     out           ; Yes
-    
-    ldy     #$00
-    sty     BGCOLOR_CHANGE
-
-    rts
-out:
-    ldx     BGCOLOR       ; Get last color in order to return it
     sta     BGCOLOR
```
[the remark this answers is not in the document]
Comment (jedeoric) [reply libsrc/telestrat/bgcolor.s]: Fixed

File: libsrc/telestrat/clrscr.s [hunk 19]
```diff
@@ -34,13 +33,14 @@
...
-    stx     BGCOLOR_CHANGE
+    
+    ; X is equal to 0
+    stx     BGCOLOR
+    stx     OLD_BGCOLOR
 
     lda     #$07
     sta     CHARCOLOR
-    sta     BGCOLOR
+    sta     OLD_CHARCOLOR
```
[the remark this answers is not in the document]
Comment (jedeoric) [reply libsrc/telestrat/clrscr.s]: Fixed

File: libsrc/telestrat/gotoxy.s (hunk 16 above)
Comment (jedeoric) [reply libsrc/telestrat/gotoxy.s]: Fixed

File: libsrc/telestrat/gotoxy.s (hunk 17 above)
Comment (jedeoric) [reply libsrc/telestrat/gotoxy.s]: Fixed

**Comment by oliverschmidt** [comment]
@greg-king5: Thanks for the effort you put into reviewing this pull request :-) Please merge it when you consider it okay.

File: libsrc/telestrat/gotox.s [hunk 20]
```diff
@@ -3,13 +3,8 @@
...
-    .import    popa
-
-    .importzp  sp
-
     .include   "telestrat.inc"
 
-
 .proc _gotox
    sta    SCRX
    rts
```
Comment (greg-king5) [inline libsrc/telestrat/gotox.s]: I think that `gotox()` needs to "reset" the old colors, as `gotoxy()` does.  Consider this program flow:
- A row is divided into three areas.
- Some text is printed in the middle area, in one color.
- Then, some text is printed in the left area, in a different color.
- Then, some text is printed in the right area, in the same color as the left area.
- `gotox()` is used to move between the areas.

Currently, no attribute is put into the right area; but, that allows the middle color "to bleed" onto the right-side text!  If 'gotox()' sets the old colors to `$FF`, then the right area will get attributes.  It always will show its desired colors.

File: libsrc/telestrat/clrscr.s [hunk 21]
```diff
@@ -30,9 +29,14 @@
...
-    lda     #$01
-    sta     SCRY
-    lda     #$00
-    sta     SCRX
+    ldx     #$01
+    stx     SCRY
+    dex
+    stx     SCRX
+    
+    lda     #$FF
```
[the remark this answers is not in the document]
Comment (jedeoric) [reply libsrc/telestrat/clrscr.s]: Fixed

File: libsrc/telestrat/clrscr.s [hunk 22]
```diff
@@ -3,24 +3,23 @@
 ;
 
     .export    _clrscr
-
-    .importzp  sp
-
+    .import    OLD_CHARCOLOR, OLD_BGCOLOR, CHARCOLOR, BGCOLOR
```
[the remark this answers is not in the document]
Comment (jedeoric) [reply libsrc/telestrat/clrscr.s]: Fixed

File: libsrc/telestrat/cputc.s [hunk 23]
```diff
@@ -4,12 +4,51 @@
...
+    ldx     BGCOLOR
+    cpx     OLD_BGCOLOR
+    beq     do_not_change_color
+
+    stx     OLD_BGCOLOR
+
+    dec     SCRX                 ; Dec SCRX in order to place attribute before the right position
+
+    pha
+    txa                          ; Swap X to A because, X contains CHARCOLOR
```
[the remark this answers is not in the document]
Comment (jedeoric) [reply libsrc/telestrat/cputc.s]: Fixed

File: libsrc/telestrat/gotox.s (hunk 20 above)
Comment (jedeoric) [reply libsrc/telestrat/gotox.s]: Fixed

**Comment by jedeoric** [comment]
Fix done @oliverschmidt i did the changed asked by @polluks in gotoy.s. In gotoxy.s, i did what @greg-king5 askes (rts removed in ".proc _gotoxy" in that case, it continues to update_adscr

File: libsrc/telestrat/clrscr.s (hunk 21 above)
[the remark this answers is not in the document]
Comment (greg-king5) [reply libsrc/telestrat/clrscr.s]: Then, the variables need to be stored from the .X index register.

**Comment by greg-king5** [comment]
I can't find anything else to improve -- merging!!

**Comment by oliverschmidt** [comment]
Thanks !

**Pull request merged**